File: content/problems/2929-distribute-candies-among-children-ii/solutions/constant-time-inclusion-exclusion/solution.cpp
class Solution {
    long long ways(long long sum){
        return sum<0?0:(sum+1)*(sum+2)/2;
    }
    public:long long distributeCandies(int n,int limit){
        long long step=limit+1LL;
        return ways(n)-3*ways(n-step)+3*ways(n-2*step)-ways(n-3*step);
    }
};

File: content/problems/2929-distribute-candies-among-children-ii/solutions/constant-time-inclusion-exclusion/Solution.kt
class Solution {
    private fun ways(sum:Long):Long=if(sum<0)0 else (sum+1)*(sum+2)/2;
    fun distributeCandies(n:Int,limit:Int):Long{
        val step=limit+1L;
        return ways(n.toLong())-3*ways(n-step)+3*ways(n-2*step)-ways(n-3*step)
    }
}
